Adding freeze argument and allowing for easier re-runs

ch-kr opened the pull request:
Major changes:

- Re-added save chi square HT option for single break search to allow for easier re-runs (avoid needing to re-run round 1 of single break search)
- Added freeze argument to pipeline and utility scripts missed last round

Minor changes:

- Added option to initialize Hail quietly (I want to test how how this looks)

ch-kr commented:
Additional changes:

- Fixed `build_models` calls (removed `trimers` arg)
- Updated from using truncated chi square value to full value (using `hl.qchisqtail`)
- Added `P_VALUE` constant
- Added new `MIN_EXP_MIS` constant and set it to 16.0
- Added `MIN_CHISQ_THRESHOLD` constant
- Changed Batch billing project from gnomad-production > rmc-production

lw453 commented on rmc/utils/simultaneous_breaks.py:
```diff
@@ -303,10 +314,11 @@ def calculate_window_chisq(
     count: int,
-    chisq_threshold: float = 9.2,
+    chisq_threshold: float = CHISQ_THRESHOLDS["simul"],
     min_num_exp_mis: float = 10,
```
Should be switched to `MIN_EXP_MIS`

lw453 commented on rmc/pipeline/regional_constraint.py:
```diff
@@ -256,25 +269,46 @@ def main(args):
-                logger.info(
-                    "Adding section annotation before searching for first break..."
+                all_loci_chisq_ht_path = (
+                    f"{SINGLE_BREAK_TEMP_PATH}/round{args.search_num}_all_loci_chisq.ht"
```
This file is currently saved as just `all_loci_chisq.ht`, do we want to change the name or just remove the `round{args.search_num}` here?

ch-kr replied:
hmmmm I'll remove ```round{args.search_num}```. I'll probably delete this HT (and regenerate since this will contain sections with min exp mis < 16

lw453 commented:
A couple of additional comments:

- Reference to `grouped_single_no_break_ht_path` in `prepare_transcripts.py` doesn't have a freeze argument passed in - should we add this?
- We can change line 231 in `rmc.py` in `single_search_round_ht_path()` to `return f"{single_search_bucket_path(search_num, freeze)}/{break_status}{breakpoint_status}.ht"`
- I think we should add the `freeze` argument to `merge_rmc_hts` and consequently `merge_round_no_break_ht` since it uses `single_search_round_ht_path` and `simul_search_round_bucket_path` (currently with the default `CURRENT_FREEZE`)

Places where we need to add freeze to a temp filename:
- Line 333 in `regional_constraint.py`: `f"{TEMP_PATH_WITH_FAST_DEL}/round{args.search_num}_temp.ht",`
- Line 547 in `regional_constraint.py`: `f"{TEMP_PATH_WITH_SLOW_DEL}/rmc_results.ht",`
- Line 522 in `constraint.py`: `f"{TEMP_PATH_WITH_FAST_DEL}/group_max_chisq.ht", overwrite=True`
- Line 641 in `constraint.py`: `f"{temp_path}/round{search_num}_all_loci_chisq.ht", overwrite=True` (if saved in `TEMP_PATH_WITH_FAST_DEL`
- Line 1095 in `constraint.py`: `f"{TEMP_PATH_WITH_FAST_DEL}/search_union.ht", overwrite=True`

ch-kr commented on rmc/resources/rmc.py:
```diff
@@ -32,15 +32,57 @@
+
+CHISQ_THRESHOLDS = {
+    "single": hl.eval(hl.qchisqtail(P_VALUE, 1)),
+    "simul": hl.qchisqtail(P_VALUE, 2),
```
nope, just forgot about it!

ch-kr commented on rmc/resources/rmc.py:
```diff
@@ -32,15 +32,57 @@
+will not be computed (chi square will be annotated as a missing value).
+
+Calculated using a power curve with code at
+https://github.com/broadinstitute/gnomad_lof/blob/master/R/efig7_constraint.R.
```
oh good to know! do you mind just adding a quick description here so we have it for future reference?

ch-kr commented:
addressed comments above and also added:

- Command line arg to toggle whether to save all loci chi square HT for single and simul break searches
- Add structure to the merged no break/break HTs (so they don't write to top level RMC temp bucket)

ch-kr commented on rmc/plotting/expected_vs_significance.R:
```diff
@@ -0,0 +1,37 @@
+    return(
+        -2 *
+            (
+                log(dpois(obs1, exp1 * (obs1 + obs2) / (exp1 + exp2))) +
```
totally small thing, but this line (and the ones below) could use the log parameter:
```
dpois(obs1, exp1 * (obs1 + obs2) / (exp1 + exp2), log.p = True) +
```

ch-kr commented on rmc/plotting/expected_vs_significance.R:
```diff
@@ -0,0 +1,37 @@
+    )
+}
+
+# Plot p-values of LRT in the above scenario, varying over
```
```suggestion
# Plot p-values of likelihood ratio test (LRT) in the above scenario, varying over
```

ch-kr reviewed (commented):
optional minor comments

lw453 commented on rmc/utils/constraint.py:
```diff
@@ -638,10 +642,11 @@ def search_for_break(
-        f"{temp_path}/round{search_num}_all_loci_chisq.ht", overwrite=True
+        f"{temp_path}/freeze{freeze}_round{search_num}_all_loci_chisq.ht",
+        overwrite=True,
     )
```
```suggestion
    all_loci_chisq_ht_path = (
        f"{TEMP_PATH_WITH_FAST_DEL}/freeze{freeze}_round{search_num}_all_loci_chisq.ht"
    )
    if save_chisq_ht:
        all_loci_chisq_ht_path = f"{SINGLE_BREAK_TEMP_PATH}/all_loci_chisq.ht"
    ht = ht.checkpoint(all_loci_chisq_ht_path, overwrite=True)

lw453 commented on rmc/utils/simultaneous_breaks.py:
```diff
@@ -458,7 +458,7 @@ def process_section_group(
         Default is False.
```
On line 398-407, we should change to:
```
    group_ht_path = (
        f"{TEMP_PATH_WITH_FAST_DEL}/freeze{freeze}_dataproc_temp_chisq_group{count}.ht"
    )
    if save_chisq_ht:
        group_ht_path = f"{SIMUL_BREAK_TEMP_PATH}/freeze{freeze}/dataproc_temp_chisq_group{count}.ht"
    ht = ht.checkpoint(all_loci_chisq_ht_path, overwrite=True)
```
Similarly for `run_batches.py` lines 373-382

ch-kr commented on rmc/resources/rmc.py:
```diff
@@ -389,12 +393,22 @@ def merged_search_ht_path(
+def no_breaks_he_path(
+    freeze: int = CURRENT_FREEZE,
+    gnomad_ver: str = CURRENT_GNOMAD_VERSION,
+    constraint_prefix: str = CONSTRAINT_PREFIX,
```
I'm fine either way, I added them as parameters just when converting from a variable to a function

lw453 commented on rmc/pipeline/two_breaks/run_batches.py:
```diff
@@ -370,16 +370,15 @@ def search_for_two_breaks(
-            f"{TEMP_PATH_WITH_FAST_DEL}/freeze{freeze}_batch_temp_chisq_group{count}.ht",
-            overwrite=True,
+        group_ht_path = (
+            f"{SIMUL_BREAK_TEMP_PATH}/freeze{freeze}_batch_temp_chisq_group{count}.ht"
```
```suggestion
            f"{SIMUL_BREAK_TEMP_PATH}/freeze{freeze}/batch_temp_chisq_group{count}.ht"
